File: ambeth/Ambeth.Cache/ambeth/cache/interceptor/CacheContextInterceptor.cs
using De.Osthus.Ambeth.Log;
using De.Osthus.Ambeth.Proxy;
using De.Osthus.Ambeth.Ioc;
#if SILVERLIGHT
using Castle.Core.Interceptor;
#else
using Castle.DynamicProxy;
#endif
using System.Reflection;
using System;
using De.Osthus.Ambeth.Util;
using De.Osthus.Ambeth.Ioc.Annotation;

namespace De.Osthus.Ambeth.Cache.Interceptor
{
    public class CacheContextInterceptor : CascadedInterceptor
    {
        [LogInstance]
		public ILogger Log { private get; set; }

        [Autowired]
        public ICacheContext CacheContext { protected get; set; }

        [Autowired]
        public ICacheFactory CacheFactory { protected get; set; }

        [Autowired]
        public ICacheProvider CacheProvider { protected get; set; }

        protected override void InterceptIntern(IInvocation invocation)
        {
            MethodInfo method = invocation.Method;
            if (method.DeclaringType.Equals(typeof(Object)))
            {
                InvokeTarget(invocation);
                return;
            }
            Object result = CacheContext.ExecuteWithCache(CacheProvider, delegate()
            {
                InvokeTarget(invocation);
                return invocation.ReturnValue;
            });
        }
    }
}

File: jambeth/jambeth-cache/src/main/java/de/osthus/ambeth/cache/interceptor/CacheContextInterceptor.java
package de.osthus.ambeth.cache.interceptor;

import java.lang.reflect.Method;

import net.sf.cglib.proxy.MethodProxy;
import de.osthus.ambeth.cache.ICacheContext;
import de.osthus.ambeth.cache.ICacheFactory;
import de.osthus.ambeth.cache.ICacheProvider;
import de.osthus.ambeth.exception.RuntimeExceptionUtil;
import de.osthus.ambeth.ioc.annotation.Autowired;
import de.osthus.ambeth.log.ILogger;
import de.osthus.ambeth.log.LogInstance;
import de.osthus.ambeth.proxy.CascadedInterceptor;
import de.osthus.ambeth.threading.IResultingBackgroundWorkerDelegate;

public class CacheContextInterceptor extends CascadedInterceptor
{
	@SuppressWarnings("unused")
	@LogInstance
	private ILogger log;

	@Autowired
	protected ICacheContext cacheContext;

	@Autowired
	protected ICacheFactory cacheFactory;

	@Autowired
	protected ICacheProvider cacheProvider;

	@Override
	protected Object interceptIntern(final Object obj, final Method method, final Object[] args, final MethodProxy proxy) throws Throwable
	{
		if (method.getDeclaringClass().equals(Object.class))
		{
			return invokeTarget(obj, method, args, proxy);
		}
		try
		{
			return cacheContext.executeWithCache(cacheProvider, new IResultingBackgroundWorkerDelegate<Object>()
			{
				@Override
				public Object invoke() throws Throwable
				{
					return invokeTarget(obj, method, args, proxy);
				}
			});
		}
		catch (Throwable e)
		{
			throw RuntimeExceptionUtil.mask(e, method.getExceptionTypes());
		}
	}
}
